File: app/src/main/java/ru/vitalydemidov/og_testapp/appcommon/adapter/BaseDelegateAdapter.java
package ru.vitalydemidov.og_testapp.appcommon.adapter;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.util.SparseArray;
import android.view.ViewGroup;
import ru.vitalydemidov.og_testapp.appcommon.model.BaseItem;

import java.util.ArrayList;
import java.util.List;

public abstract class BaseDelegateAdapter<VM extends BaseItem> extends RecyclerView.Adapter<AbstractBaseItemViewHolder<VM>> {

    private SparseArray<BaseViewTypeDelegate> delegateMap = new SparseArray<>();

    private List<BaseItem> dataList = new ArrayList<>();

    public void setDataList(List<BaseItem> list) {
        dataList = list;
        notifyDataSetChanged();
    }

    public void addViewTypeDelegate(int itemType, @NonNull BaseViewTypeDelegate delegate) {
        delegateMap.append(itemType, delegate);
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public AbstractBaseItemViewHolder<VM> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        BaseViewTypeDelegate delegate = getViewTypeDelegate(viewType);
        if (delegate != null) {
            return delegate.onCreateViewHolder(parent, viewType);
        } else {
            throw new IllegalArgumentException("Invalid viewType");
        }
    }

    @SuppressWarnings("unchecked")
    @Override
    public void onBindViewHolder(@NonNull AbstractBaseItemViewHolder<VM> holder, int position) {
        BaseViewTypeDelegate delegate = getViewTypeDelegate(holder.getItemViewType());
        VM item = (VM) dataList.get(position);
        if (delegate != null && item != null) {
            delegate.onBindViewHolder(holder, item);
        }
    }

    @Nullable
    private BaseViewTypeDelegate getViewTypeDelegate(int itemType) {
        return delegateMap.get(itemType);
    }

    @Override
    public int getItemCount() {
        return dataList.size();
    }

    @Override
    public int getItemViewType(int position) {
        return dataList.get(position).getType();
    }

}

File: app/src/main/java/ru/vitalydemidov/og_testapp/appcommon/adapter/BaseDelegateAdapter.kt
package ru.vitalydemidov.og_testapp.appcommon.adapter

import android.support.v4.util.SparseArrayCompat
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import ru.vitalydemidov.og_testapp.appcommon.model.BaseItem

abstract class BaseDelegateAdapter<VM> :
    RecyclerView.Adapter<AbstractBaseItemViewHolder<VM>>() {

    private val viewTypeDelegates = SparseArrayCompat<BaseViewTypeDelegate<VM, AbstractBaseItemViewHolder<VM>>>()

    internal var dataList: List<BaseItem<VM>> = arrayListOf()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    protected fun addViewTypeDelegate(
        itemType: Int,
        viewTypeDelegate: BaseViewTypeDelegate<VM, AbstractBaseItemViewHolder<VM>>
    ) {
        viewTypeDelegates.append(itemType, viewTypeDelegate)
    }

    internal fun getViewTypeDelegate(itemType: Int): BaseViewTypeDelegate<VM, AbstractBaseItemViewHolder<VM>>? =
        viewTypeDelegates.get(itemType)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AbstractBaseItemViewHolder<VM> {
        val viewTypeDelegate = getViewTypeDelegate(viewType)
        return viewTypeDelegate?.onCreateViewHolder(parent, viewType)
            ?: throw IllegalArgumentException("Invalid viewType")
    }

    override fun onBindViewHolder(holder: AbstractBaseItemViewHolder<VM>, position: Int) {
        val viewTypeDelegate = getViewTypeDelegate(holder.itemViewType)
        val item = dataList[position]
        viewTypeDelegate?.onBindViewHolder(holder, item)
    }

    override fun getItemCount(): Int = dataList.size

}